Fix handling of missing values in median and quantile and improve docs

nalimilan opened the pull request:
Ensure we always return missing or throw an error in the presence of missing values
Support `skipmissing(v)` by supporting any iterator and calling collect on it, which makes `median` and `quantile` consistent with `mean` and `sum`.
Improve docstrings for stats functions by adding examples and making them more consistent.

Fixes https://github.com/JuliaLang/julia/issues/27363.

nalimilan commented on base/statistics.jl:
```diff
@@ -177,29 +241,48 @@ _median(v::AbstractArray{T}, ::Colon) where {T} = median!(copyto!(Array{T,1}(und
-    throw an `ArgumentError` in the presence of `NaN` values in the data array.
-    Use the [`missing`](@ref) type to represent missing values, and the
-    [`skipmissing`](@ref) function to omit them.
+    An `ArgumentError` is thrown if `v` contains `NaN` or [`missing`](@ref) values.
```
I have preserved the current behavior, which is inconsistent with `median` (and `mean`, `sum`, etc.). I can change it now to return `missing` instead of throwing an error, but we can do that at any time in the 1.x series since that's not breaking.

andreasnoack replied:
It seems unfortunate if `median` and `quantile` don't have the same behavior wrt `missing`s

nalimilan replied:
I'm fine with changing `quantile` to return `missing` instead of throwing an error, but that requires a bit more work/thought given that `quantile` returns an array when passed several quantile levels, and yet we probably don't want to make it an `Array{Union{T,Missing}}`. So I'd rather change this in another PR so that we can discuss it more thoroughly.

andreasnoack replied:
Fine with me

nalimilan replied:
Filed https://github.com/JuliaLang/julia/issues/27905 so that we don't forget.

nalimilan commented on base/statistics.jl:
```diff
@@ -139,11 +176,8 @@ Like [`median`](@ref), but may overwrite the input vector.
-        end
-    end
+    eltype(v)>:Missing && any(ismissing, v) && return missing
+    eltype(v)>:AbstractFloat && any(isnan, v) && return NaN
```
I broke this check by mistake but interestingly the tests didn't catch it because they only covered cases like `[1.0, NaN]`, which return `NaN` without it. The correct check seems to be `eltype(v)<:AbstractFloat || eltype(v)>:AbstractFloat`, right?

Ideally, we wouldn't need these explicit type checks (https://github.com/JuliaLang/julia/issues/21256).

andreasnoack replied:
What if the input is say `Any[1, missing, 1]` or maybe even `Any[1, missing, NaN, 1]`?

nalimilan replied:
Since `Any >: Missing`, these return `missing`, as documented in the docstring (below).

nalimilan replied:
Regarding the other point, do you think `eltype(v)<:AbstractFloat || eltype(v)>:AbstractFloat` is OK here?

nalimilan commented:
Bump.

andreasnoack commented on base/statistics.jl:
```diff
@@ -66,13 +90,26 @@ function mean!(R::AbstractArray, A::AbstractArray)
+    mean(A::AbstractArray; dims)
 
-Compute the mean of whole array `v`, or optionally along the given dimensions.
+Compute the mean of an array over the given dimensions.
```
I'm not sure this reformulation is an improvement

nalimilan replied:
Note that I've added a separate docstring (above) for the simpler method which doesn't take `dims` and isn't restricted to `AbstractArray`. This is consistent with other reductions. What would you like to change?

andreasnoack commented on base/statistics.jl:
```diff
@@ -177,29 +241,48 @@ _median(v::AbstractArray{T}, ::Colon) where {T} = median!(copyto!(Array{T,1}(und
 
-Compute the quantile(s) of a vector `v` at the probability or probabilities `p`, which
-can be given as a single value, a vector, or a tuple. If `p` is a vector, an optional
+Compute the quantile(s) of vector `v` at a specified probability or vector or tuple of
```
I think you'd still want *a* or *the* vector and I'm not sure the reformulation is an improvement.

nalimilan replied:
OK, I've just tried to use the same wording for `quantile` and `quantile!`. Can you suggest something else?

andreasnoack replied:
The comment included two proposals: either keep the old formulation as it is or just add "a" or "the" before "vector".

andreasnoack commented on base/statistics.jl:
```diff
@@ -273,26 +357,40 @@ end
-be sorted.
-
-The `p` should be on the interval [0,1], and `v` should not have any `NaN` values.
+Compute the quantile(s) of collection `itr` at a specified probability or vector or tuple of
```
*a* or *the* collection

nalimilan commented:
I've fixed the remaining issues and added some tests. I'll merge in a while barring objections, but feel free to comment on the question at https://github.com/JuliaLang/julia/pull/27515#discussion_r197804372...

nalimilan committed:
Fix handling of missing values in median and quantile and improve docs

Ensure we always return missing or throw an error in the presence of missing values
Support skipmissing(v) by supporting any iterator and calling collect on it, which
makes median and quantile consistent with mean and sum.
Improve docstrings for stats functions by adding examples and making them more consistent.

nalimilan committed:
Improve docstrings and tests for std[m], var[m] about missing values
